### DISCUSSION TITLE
export result while equations are evolving

### DISCUSSION THREAD

**Discussion by AKHCE** [MilesCranmer/PySR#439] [Q&A] [answered]
Hi,
By passing equation_file and/or temp_equation_file to PySRRegressor, we can save the latest updated "hall of fame". For example,
![intermediate](https://github.com/MilesCranmer/PySR/assets/122554992/759b107d-77b7-4fba-aaaa-86228e311a7d)
The results can be saved as csv.

The output csv file gets updated while evolution. In other words, we always have one csv file which is being updated. I am wondering if there is a way to keep all intermediate csv files (or at least some of them). For example, can we store 1 separated csv file per niterations with its own unique name? What I want to finally achieve is to plot loss versus evolution for each Complexity. e.g. complexity = 5, then plotting the corresponding loss versus iterations/generations. 

Any advice is much appreciated, 

Thank you so much for your support,
Cheers

**Comment by MilesCranmer** [comment]
Hi @AKHCE,

Sorry for the late reply, I missed this post for some reason.

There is not a built-in mechanism for saving the csv file over time. However, there is a mechanism for saving the entire history of a search to a JSON file (including every single mutation that happens). It is a bit overkill and I'm not sure if it is useful, but you could try it out. 

It's typically used from the Julia side so there's no built-in parameter in PySRRegressor. However, you could turn this on by adding to:

https://github.com/MilesCranmer/PySR/blob/6abcc15afe7953b08b4f15d1a3e9a127e4e40691/pysr/sr.py#L1608

add a line

```python
val_recorder=Main.eval("Val(true)")
```

and it should generate a JSON file `pysr_recorder.json` of the history of the search. Just keep in mind that it is going to be much slower. This JSON file is used only as a debugging tool (by myself) so there's not documentation on it.

Other than that, you could modify the backend, particularly [this line](https://github.com/MilesCranmer/SymbolicRegression.jl/blob/4f0b8c576eaccbdcbb6497ab446a3abf6535aaec/src/SymbolicRegression.jl#L898), and get it to save a file given by the current `time()` rounded to some precision. 

You can use a custom backend in PySR by following the instructions here: https://astroautomata.com/PySR/backend/

Cheers,
Miles

**Reply by AKHCE** [reply, marked as answer]
Thank you so much for giving me alternatives. I chose a different approach which seems working well. Simply, I pass warm_start = True to enable resuming the training, then I use a for-loop to move forward the model training. Please see screenshot below:
![image](https://github.com/MilesCranmer/PySR/assets/122554992/84b1a93a-92ab-4b0e-9c48-059f90c6f72d)
As you can see I set populations (I interpret it as generation) to 1 and the model is trained through a for-loop. Also, niterations which is number of re-running algorithm in each generation can be set any values, I chose 1 in the given example.

As you can see, I can save hall of fame for each generation. How do you think about this approach?
Thank you so much for your support,

**Reply by MilesCranmer** [reply]
Oh yeah, that’s way easier!!! Sorry for my overcomplicated approach, I’ve done so much dev on the internals that I forgot there was a simple approach like this!

**Reply by MilesCranmer** [reply]
Btw, when using `pow`, you should turn on `constraints={"pow": (-1, 1)}`, otherwise it will just start stacking power laws everywhere.

**Reply by AKHCE** [reply]
Absolutely. Thanks
